### PR TITLE
[inductor][overlap] fix circular extra deps; handle artifactos of collapsing fused regions 

### PR DESCRIPTION
**Pull request by IvanKobzarev**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #172081
Fixes CI failure in autoparallel torchtitan CI: https://github.com/meta-pytorch/autoparallel/actions/runs/20448109712/job/58755596788?pr=291#logs

Repro:

```
NGPU=4 CONFIG_FILE="./torchtitan/models/llama3/train_configs/debug_model.toml" ./run_train.sh   --model.name autoparallel.llama3   --parallelism.tensor_parallel_degree 4   --training.dataset c4   --compile.enable   --job.custom_config_module=torchtitan.experiments.autoparallel.job_config
```

There are 2 independent problems:
1. During bucketing transformations when replaced nodes dependencies are applied to augment_graph_helper, it was adding recursive dep (_pre_bucket_collective -> _pre_bucket_collective) which resulted in failure of stable topological sort with additional deps.

2. If fused regions are enabled (enabled by default) - extra deps were applied to augment_graph_helper, some of them were erased during fused region expand (get_item of result of call_module when result is tuple) and failed on assert that additional deps have been erased.

3. There is one more problem in additional deps lowering in scheduler.py, where buffer name with dep name ("primals_14" in failure) was not found.

cc @ezyang @EikanWang @jgong5 @wenzhe-nrv @voznesenskym @penguinwu @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

File: torch/_inductor/fx_passes/fusion_regions.py [hunk 1]
```diff
@@ -311,6 +312,11 @@ def expand_fusion_regions(
             f"Expected submodule {subgraph_name} to exist"
         )
 
+        # Users of module_node are get_items that will be removed from the graph
+        for user in module_node.users:
+            if user.op == "call_function" and user.target == operator.getitem:
```
Comment (Skylion007) [inline torch/_inductor/fx_passes/fusion_regions.py:317]:
```suggestion
            if user.op == "call_function" and user.target is operator.getitem:

File: torch/_inductor/fx_passes/fusion_regions.py [hunk 2]
```diff
@@ -311,6 +312,11 @@ def expand_fusion_regions(
...
+                result[user] = None
```
Comment (eellison) [inline torch/_inductor/fx_passes/fusion_regions.py:318]: nit: can we resolve removing this here, instead of in augmented graph helper?

File: torch/_inductor/scheduler.py [hunk 3]
```diff
@@ -688,6 +688,9 @@ def prune_weak_deps(self) -> None:
         def should_prune(dep: Dep) -> bool:
             if not isinstance(dep, WeakDep):
                 return False
+            if dep.name not in self.scheduler.name_to_buf:
```
Comment (eellison) [inline torch/_inductor/scheduler.py:725]: Could we check that this is an input buffer, and return False, to handle this case more explicitly ? that was this is fixing. There's a view etc of an input buffer and then a dep gets added.

File: torch/_inductor/fx_passes/fusion_regions.py [hunk 4]
```diff
@@ -287,7 +288,7 @@ def collapse_fusion_regions(
 def expand_fusion_regions(
     gm: fx.GraphModule,
     region_of: dict[fx.Node, FusionRegion],
-) -> dict[fx.Node, fx.Node]:
+) -> dict[fx.Node, fx.Node | None]:
```
Comment (eellison) [inline torch/_inductor/fx_passes/fusion_regions.py:291]: Looking at the repro, claude came up with these two tests. Can you add them to bucketing_unit ?:
```

@requires_accelerator_dist_backend(["nccl", "xccl"])
@unittest.skipIf(not HAS_GPU, "Inductor+gpu needs triton and recent GPU arch")
class TestOverlapSchedulingFixes(InductorTestCase):
    """
    Test cases for specific bug fixes in overlap scheduling.
    These tests would fail without their corresponding fixes.
    """

    @classmethod
    def setUpClass(cls):
        super().setUpClass()
        from torch.testing._internal.distributed.fake_pg import FakeStore

        store = FakeStore()
        dist.init_process_group(backend="fake", rank=0, world_size=16, store=store)
        cls.device = "cuda"

    @classmethod
    def tearDownClass(cls):
        super().tearDownClass()
        dist.destroy_process_group()

    def test_no_self_dependency_cycle_with_dtype_conversion(self):
        """
        Test that bucketing collectives with dtype conversion doesn't create
        self-dependency cycles.

        This tests the fix in augmented_graph_helper.py that adds != new_node
        checks to prevent self-dependencies when merging nodes.

        The bug: When two convert_element_type nodes (inputs to all_gathers)
        have timeline dependencies between them and both get merged into
        _pre_bucket_all_gather, the dependency becomes a self-dependency
        which causes _stable_topological_sort to fail.
        """

        def func(a, b, c, d):
            group_name = dist.distributed_c10d._get_default_group().group_name
            group_size = 16

            # Multiple all_gathers with dtype conversion
            # The convert nodes will have timeline dependencies between them
            conv_a = torch.ops.prims.convert_element_type.default(a, torch.bfloat16)
            ag1 = torch.ops._c10d_functional.all_gather_into_tensor(
                conv_a, group_size, group_name
            )

            conv_b = torch.ops.prims.convert_element_type.default(b, torch.bfloat16)
            ag2 = torch.ops._c10d_functional.all_gather_into_tensor(
                conv_b, group_size, group_name
            )

            # Compute between all_gathers
            mm = torch.mm(c, d)

            conv_c = torch.ops.prims.convert_element_type.default(c, torch.bfloat16)
            ag3 = torch.ops._c10d_functional.all_gather_into_tensor(
                conv_c, group_size, group_name
            )

            conv_d = torch.ops.prims.convert_element_type.default(d, torch.bfloat16)
            ag4 = torch.ops._c10d_functional.all_gather_into_tensor(
                conv_d, group_size, group_name
            )

            # Wait for all
            w1 = torch.ops._c10d_functional.wait_tensor(ag1)
            w2 = torch.ops._c10d_functional.wait_tensor(ag2)
            w3 = torch.ops._c10d_functional.wait_tensor(ag3)
            w4 = torch.ops._c10d_functional.wait_tensor(ag4)

            return w1.sum() + w2.sum() + w3.sum() + w4.sum() + mm.sum()

        with FakeTensorMode():
            a = torch.ones(4, 4, device=self.device)
            b = torch.ones(4, 4, device=self.device)
            c = torch.ones(4, 4, device=self.device)
            d = torch.ones(4, 4, device=self.device)

            traced = make_fx(func)(a, b, c, d)

        # Run full overlap scheduling with bucketing enabled
        # This would fail with AssertionError in _stable_topological_sort
        # before the self-dependency fix
        from torch._inductor.fx_passes.overlap_scheduling import OverlapScheduler

        scheduler = OverlapScheduler(
            traced,
            max_in_flight_gb=5.0,
            max_compute_pre_fetch=200,
            collective_bucketing=True,
            insert_overlap_deps=True,
            compute_overlap_multipler=1.0,
            max_coll_distance=200,
            custom_runtime_estimation=None,
            collective_estimator="analytical",
            enable_fusion_regions=False,
        )
        # This should complete without cycle error
        result = scheduler.run()
        result.graph.lint()

    def test_no_cycle_with_fusion_regions_and_bucketing(self):
        """
        Test that fusion regions + bucketing doesn't create cycles.

        This tests multiple fixes:
        1. Self-dependency prevention (augmented_graph_helper.py)
        2. Track erased getitem nodes (const_fold.py, fusion_regions.py)
        3. Skip DCE during expansion (const_fold.py, fusion_regions.py)

        The scenario: Fusion regions collapse fusible ops into call_module nodes.
        When bucketing merges collectives, getitem nodes from fusion outputs
        get erased. Without proper tracking and DCE skip, this causes cycles
        or assertion failures.
        """

        def func(a, b, c, d):
            group_name = dist.distributed_c10d._get_default_group().group_name
            group_size = 16

            # Start collectives
            conv_a = torch.ops.prims.convert_element_type.default(a, torch.bfloat16)
            ag1 = torch.ops._c10d_functional.all_gather_into_tensor(
                conv_a, group_size, group_name
            )

            conv_b = torch.ops.prims.convert_element_type.default(b, torch.bfloat16)
            ag2 = torch.ops._c10d_functional.all_gather_into_tensor(
                conv_b, group_size, group_name
            )

            # Fusible compute chain (will become a fusion region)
            x = c + 1
            x = x * 2
            x = x - 3
            x = x / 4

            # Wait and use results
            w1 = torch.ops._c10d_functional.wait_tensor(ag1)
            w2 = torch.ops._c10d_functional.wait_tensor(ag2)

            # More collectives
            conv_c = torch.ops.prims.convert_element_type.default(c, torch.bfloat16)
            ag3 = torch.ops._c10d_functional.all_gather_into_tensor(
                conv_c, group_size, group_name
            )

            conv_d = torch.ops.prims.convert_element_type.default(d, torch.bfloat16)
            ag4 = torch.ops._c10d_functional.all_gather_into_tensor(
                conv_d, group_size, group_name
            )

            # Another fusible chain
            y = d + 1
            y = y * 2

            w3 = torch.ops._c10d_functional.wait_tensor(ag3)
            w4 = torch.ops._c10d_functional.wait_tensor(ag4)

            return w1.sum() + w2.sum() + w3.sum() + w4.sum() + x.sum() + y.sum()

        with FakeTensorMode():
            a = torch.ones(4, 4, device=self.device)
            b = torch.ones(4, 4, device=self.device)
            c = torch.ones(4, 4, device=self.device)
            d = torch.ones(4, 4, device=self.device)

            traced = make_fx(func)(a, b, c, d)

        # Run with fusion regions enabled - this exercises all the fixes
        from torch._inductor.fx_passes.overlap_scheduling import OverlapScheduler

        scheduler = OverlapScheduler(
            traced,
            max_in_flight_gb=5.0,
            max_compute_pre_fetch=200,
            collective_bucketing=True,
            insert_overlap_deps=True,
            compute_overlap_multipler=1.0,
            max_coll_distance=200,
            custom_runtime_estimation=None,
            collective_estimator="analytical",
            enable_fusion_regions=True,  # Enable fusion regions
        )
        # This should complete without errors
        result = scheduler.run()
        result.graph.lint()
```

**Review by eellison — APPROVED** [review]

**Review by sanketpurandare — APPROVED** [review]

File: test/distributed/test_overlap_bucketing_unit.py [hunk 5]
```diff
@@ -1017,5 +1017,195 @@ def func(a):
...
+                conv_a, group_size, group_name
+            )
+
+            conv_b = torch.ops.prims.convert_element_type.default(b, torch.bfloat16)
+            ag2 = torch.ops._c10d_functional.all_gather_into_tensor(
+                conv_b, group_size, group_name
+            )
+
+            # Compute between all_gathers
+            mm = torch.mm(c, d)
```
Comment (bdhirsh) [inline test/distributed/test_overlap_bucketing_unit.py:1073]: just checking - `c` and `d` here are the original inputs prior to running allgather. did you mean to run the mm on `ag1, ag2`?
